### PR TITLE
Refactor and clean up iterative solver.

### PR DESCRIPTION
**Pull request by anirudhSK**

### CODE REVIEW

File: chipc/iterative_solver.py [hunk 1]
```diff
@@ -38,81 +33,37 @@ def main(argv):
...
-    # Generate the result file
-    with open("/tmp/" + sketch_name + "_result.holes", "w") as result_file:
-        result_file.write(output)
     # Step2: run sol_verify.py
-    ret_code = sol_verify(sketch_name + "_codegen.sk",
-                          "/tmp/" + sketch_name + "_result.holes")
-    # ret_code = 0 means sol_verify success
+    ret_code = compiler.sol_verify(hole_assignments = hole_assignments, num_input_bits = 10)
     if ret_code == 0:
         print("success")
```
Comment (taegyunkim) [inline chipc/iterative_solver.py]:
```suggestion
         print("Successfully verified hole value assignments from 2 bit inputs with 10 bit inputs.")
```

File: chipc/iterative_solver.py [hunk 2]
```diff
@@ -38,81 +33,37 @@ def main(argv):
...
-                original_sketch_file_string[0:begin_pos] + "assert(!(" +
-                hole_value_file_string + "));\n" +
-                original_sketch_file_string[begin_pos:])
+            # hole_assignments is in the format {'hole_name':'hole_value'},
+            # i.e., {'sample1_stateless_alu_0_0_mux1_ctrl': '0'}
+            hole_elimination_assert = "!" # The ! is to ensure a certain hole combination isn't present.
+            for hole, value in hole_assignments.items():
+                hole_elimination_assert += "(" + hole + " == " + value + ") && "
+            hole_elimination_assert += "1"
+            (ret_code1, output, _) = compiler.serial_codegen(additional_constraints = [hole_elimination_assert])
```
Comment (taegyunkim) [inline chipc/iterative_solver.py]: Is this intentional not using hole_asignments returned from here?

File: chipc/iterative_solver.py [hunk 3]
```diff
@@ -38,81 +33,37 @@ def main(argv):
...
         print("success")
-        end_time = time.time()
-        print("total time in seconds: ", end_time - start_time)
         return 0
 
     print("failed for larger size and need repeated testing by sketch")
     # start to repeated run sketch until get the final result
-    original_sketch_file_string = Path(sketch_name + "_codegen.sk").read_text()
     count = 0
     while 1:
```
Comment (taegyunkim) [inline chipc/iterative_solver.py:100]: It's little unclear to me why we need this while loop. @XiangyuG can you remind me again? and can we add his explanation as a comment?

File: chipc/iterative_solver.py [hunk 4]
```diff
@@ -38,81 +33,37 @@ def main(argv):
...
         if mode == "hole_elimination_mode":
```
Comment (taegyunkim) [inline chipc/iterative_solver.py:101]: What do you think about splitting this if and else into two functions?

File: chipc/iterative_solver.py [hunk 5]
```diff
@@ -152,45 +103,22 @@ def main(argv):
...
+            ret_code = compiler.sol_verify(hole_assignments, 10)
             if ret_code == 0:
                 print("finally succeed")
-                end_time = time.time()
-                print("total time in seconds: ", end_time - start_time)
                 return 0
             else:
                 count = count + 1
                 continue
         else:
```
Comment (taegyunkim) [inline chipc/iterative_solver.py:123]: It's unclear to me what it means to reach this state.

File: chipc/sketch_generator.py [hunk 6]
```diff
@@ -193,10 +202,14 @@ def generate_alus(self):
...
-        template = (self.jinja2_env_.get_template("code_generator.j2")
-                    if mode == "codegen" else
-                    self.jinja2_env_.get_template("sketch_functions.j2"))
+    def generate_sketch(self, program_file, mode, additional_constraints = [], hole_assignments = dict(),
+                        additional_testcases = "", input_offset = 0):
+        self.reset_holes_and_asserts()
+        if mode == "codegen" or mode == "sol_verify" or mode == "cexgen":
+            template = self.jinja2_env_.get_template("code_generator.j2")   # TODO: Need better name for j2 file.
+        else:
+            assert(mode == "optverify"), "Found mode " + mode
```
Comment (taegyunkim) [inline chipc/sketch_generator.py]: Just a note, we should use enums instead of these hard coded strings.

**Review by taegyunkim — CHANGES_REQUESTED** [review]

File: chipc/iterative_solver.py (hunk 2 above)
Comment (anirudhSK) [reply chipc/iterative_solver.py]: Good catch. That's a bug, which we don't have a test case for. I'll fix this and the other suggestions below. Then both of you can look through it. In particular, @XiangyuG should check the code for correctness.

File: chipc/iterative_solver.py (hunk 4 above)
Comment (XiangyuG) [reply chipc/iterative_solver.py:101]: Oh yeah. It is a good idea to generate two function separately for hole_elimination_mode and cex_mode

File: chipc/iterative_solver.py (hunk 5 above)
Comment (XiangyuG) [reply chipc/iterative_solver.py:123]: I just want to know how long it takes to compile the domino program with our alu. Sometimes, it take over 48 hours to get the result.

File: chipc/iterative_solver.py (hunk 2 above)
Comment (anirudhSK) [reply chipc/iterative_solver.py]: This should be fixed with commits b9255a3 through 0ec2e47.

File: chipc/iterative_solver.py (hunk 3 above)
Comment (anirudhSK) [reply chipc/iterative_solver.py:100]: There's now a comment on top of the while loop.

File: chipc/iterative_solver.py (hunk 4 above)
Comment (anirudhSK) [reply chipc/iterative_solver.py:101]: Done.

File: chipc/sketch_generator.py (hunk 6 above)
Comment (anirudhSK) [reply chipc/sketch_generator.py]: Done.

File: chipc/iterative_solver.py (hunk 5 above)
Comment (anirudhSK) [reply chipc/iterative_solver.py:123]: I have removed timing-related code here. We can use the UNIX time utilities to measure time.

**Review by XiangyuG — COMMENTED** [review]
I find there is one more space?

**Review by XiangyuG — APPROVED** [review]

**Review by XiangyuG — APPROVED** [review]

**Review by taegyunkim — APPROVED** [review]

**Pull request merged**